File: frontend/src/api/credit-report/list.ts
/**
 * `GET /credit-report` — backend-driven paginated Credit Report list.
 *
 * All filtering, sorting, and pagination is server-driven — do NOT apply any
 * client-side transforms to the returned rows.
 *
 * `buildQuery` strips `undefined` and `""` (and we never forward `null`), so
 * every optional filter is forwarded only when truthy.
 */

import { buildQuery, getList } from "@/api/client"
import type { PaginatedResult } from "@/types/api/envelope"
import type { CreditReport, CreditReportListQuery } from "@/types/credit-report/credit-report"

/**
 * Normalise `CreditReportListQuery` into a plain record safe for `buildQuery`.
 * Keys are snake_case to match the backend `CreditReportListQuery` exactly.
 */
function buildParams(
  q: CreditReportListQuery,
): Record<string, string | number | undefined> {
  return {
    page: q.page,
    limit: q.limit,
    sortBy: q.sortBy,
    sortOrder: q.sortOrder,
    // single report-date filter ("dd-MM-yyyy", exact match on report_date)
    ...(q.date                ? { date: q.date }                               : {}),
    // 4 per-field exact-match async-select filters — only include when non-empty
    ...(q.customer_name       ? { customer_name: q.customer_name }             : {}),
    ...(q.city                ? { city: q.city }                               : {}),
    ...(q.customer            ? { customer: q.customer }                       : {}),
    ...(q.cca_description     ? { cca_description: q.cca_description }         : {}),
    // 2 enum filters — only include when non-empty
    ...(q.blocked             ? { blocked: q.blocked }                         : {}),
    ...(q.credit_balance_sign ? { credit_balance_sign: q.credit_balance_sign } : {}),
  }
}

/**
 * Fetch a paginated, server-filtered page of Credit Report rows.
 *
 * @param query - Optional query parameters; defaults to page 1, limit 20,
 *                sorted by `created_at` descending with no active filters.
 * @returns A paginated result containing the matching `CreditReport` rows
 *          and standard `PaginationMeta` (total, page, limit, totalPages).
 */
export function listCreditReport(
  query: CreditReportListQuery = {},
): Promise<PaginatedResult<CreditReport>> {
  return getList<CreditReport>(`/credit-report${buildQuery(buildParams(query))}`)
}

File: backend/app/services/credit_report/list.py
# app/services/credit_report/list.py
"""Credit Report domain business logic: backend-driven paginated listing.

Mirrors ``services/customer_code/list.py`` exactly — same pagination pattern,
same return-signature shape ``tuple[list[DTO], PaginationMeta]``.
All filtering and sorting run in the DB layer; no Python-side post-filtering
(``backend-api-standards``, no client-side filtering).

No region/FK lookup needed — customer_name is a native column stored directly.
"""
from __future__ import annotations

from math import ceil

from ...core.responses import PaginationMeta
from ...models.credit_report import CreditReport
from ...schemas.credit_report import CreditReportListQuery
from ...schemas.credit_report_record import CreditReportPublic
from ...utils.credit_report.query import build_credit_report_filter, build_sort
from .serialize import to_credit_report_public


async def list_credit_report(
    query: CreditReportListQuery,
) -> tuple[list[CreditReportPublic], PaginationMeta]:
    """Return a page of credit report rows plus stable pagination metadata.

    Filtering, sorting, and pagination all run in the DB layer.

    Args:
        query: Validated list-query DTO carrying page, limit, sortBy,
               sortOrder, the optional ``date`` filter, 4 per-field
               async-select filters (customer_name, city, customer,
               cca_description), and 2 enum filters (blocked,
               credit_balance_sign).

    Returns:
        A 2-tuple of ``(items, meta)`` where ``items`` is the serialised
        page and ``meta`` carries pagination bookkeeping for the frontend.
    """
    filt = build_credit_report_filter(query)

    total = await CreditReport.find(filt).count()

    docs = (
        await CreditReport.find(filt)
        .sort(build_sort(query.sortBy, query.sortOrder))
        .skip(query.skip)
        .limit(query.limit)
        .to_list()
    )

    items = [to_credit_report_public(doc) for doc in docs]

    meta = PaginationMeta(
        page=query.page,
        limit=query.limit,
        total=total,
        totalPages=ceil(total / query.limit) if query.limit else 0,
        sortBy=query.sortBy,
        sortOrder=query.sortOrder,
    )

    return items, meta
